storage/mega/storage_mega.go

```diff
@@ -0,0 +1,277 @@
+}
+
+// Protocols returns the Protocol Schemes supported by this backend
+func (backend *StorageMega) Protocols() []string {
+	return []string{"mega"}
+}
+
+// Description returns a user-friendly description for this backend
+func (backend *StorageMega) Description() string {
+	return "mega.nz storage"
+}
+
+// AvailableSpace returns the free space on this backend
+func (backend *StorageMega) AvailableSpace() (uint64, error) {
+	quota, err := backend.mega.GetQuota()
+	if err != nil {
+		return 0, knoxite.ErrAvailableSpaceUnknown
+	}
+	return quota.Mstrg, nil
+}
+
+// CreatePath creates a dir including all its parent dirs, when required
+func (backend *StorageMega) CreatePath(path string) error {
+	path = strings.TrimPrefix(path, "/")
+	path = strings.TrimSuffix(path, "/")
+	slicedPath := strings.Split(path, "/")
+	currentRoot := backend.mega.FS.GetRoot()
+
+	for _, pathSlice := range slicedPath {
+		// get all nodes in current root directory
+		nodesInCurrentRoot, err := backend.mega.FS.PathLookup(currentRoot, []string{pathSlice})
+		if err != nil {
+			// if we couldn't find a node at this path we need to create it
+			currentRoot, err = backend.mega.CreateDir(pathSlice, currentRoot)
+			if err != nil {
+				return err
+			}
+		} else {
+			// if nodes with the same name exist we take the node at index 0
+			currentRoot = nodesInCurrentRoot[0]
+		}
+	}
+	return nil
+}
+
+// Stat returns the size of a file
+func (backend *StorageMega) Stat(path string) (uint64, error) {
+	node, err := backend.GetLastNodeInPath(path)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(node.GetSize()), nil
+}
+
+// ReadFile reads a file from mega
+func (backend *StorageMega) ReadFile(path string) ([]byte, error) {
+
+	nodeToRead, err := backend.GetLastNodeInPath(path)
+	if err != nil {
+		return nil, err
+	}
+
+	download, err := backend.mega.NewDownload(&nodeToRead)
+	if err != nil {
+		return nil, err
+	}
+	var bytes []byte
+
+	for i := 0; i < download.Chunks(); i++ {
+		chunkBytes, err := download.DownloadChunk(i)
+		if err != nil {
+			return nil, err
+		}
+		bytes = append(bytes, chunkBytes...)
+	}
+
+	download.Finish()
+	return bytes, nil
+}
+
+// WriteFile writes file on mega
+func (backend *StorageMega) WriteFile(path string, data []byte) (size uint64, err error) {
+	var progress chan int
+	dir, file := filepath.Split(path)
+
+	_, err = backend.GetLastNodeInPath(path)
+	if err == nil {
+		// sadly, if the file exists it needs to be deleted before re-uploading, otherwise there will be a copy
+		err = backend.DeleteFile(path)
+		if err != nil {
+			return 0, err
+		}
+	}
+	nodeToWriteIn, err := backend.GetLastNodeInPath(dir)
+	if err != nil {
+		return 0, err
+	}
+	upload, err := backend.mega.NewUpload(&nodeToWriteIn, file, int64(len(data)))
+	if err != nil {
+		return 0, err
+	}
+
+	// following code is based on https://sourcegraph.com/github.com/t3rm1n4l/go-mega@79a813bb328ddee4889f53a97fbdcd71e5ae1a09/-/blob/mega.go#L1532:16
+	workch := make(chan int)
```

Comment:
After some pondering, I think we should just try to use a single connection/thread to upload the file (like we do for the downloads). Not only does it keep the code a bit neater, knoxite itself already tries to multiplex connections to make the most of the available bandwidth, so this might be counter-productive.